junebug/logging_service.py

```diff
@@ -122,3 +124,72 @@ def stopService(self):
 
     def registered(self):
         return self.log_observer in self.logger.observers
+
+
+def reverse_read(filename, buf):
+    '''
+    Read the non-blank lines from a file in reverse order.
+
+    :param str filename: The path of the file to be read.
+    :param int buf: The size of the read buffer.
+
+    :returns:
+        A generator that yields each line from the file in reverse order as a
+        string.
+    '''
+    with open(filename) as f:
+        f.seek(0, os.SEEK_END)
+        total_size = remaining_size = f.tell()
+        offset = 0
+        incomplete_line = None
+        while remaining_size > 0:
+            offset = min(total_size, offset + buf)
+            f.seek(total_size - offset, os.SEEK_SET)
+            data = f.read(min(remaining_size, buf))
+            # If we read no data, we should exit the loop
+            if len(data) == 0:
```

Comment:
We should be okay if `f.read()` returns fewer bytes than requested, since we're using `len(data)` for everything now.

I can see how we can simplify by removing `total_size` and `offset`, so I'll remove those.